cmd/provisioner/main.go

```diff
@@ -17,103 +17,46 @@
-	config.Version = sarama.V0_11_0_0
-	config.ClientID = "kafka-provisioner"
-	admin, err := sarama.NewClusterAdmin([]string{broker}, config)
+func handleCreationRequest(broker, gateway string, writer http.ResponseWriter, request *http.Request) {
+	kafkaClient, err := client.NewKafkaClient(broker, os.Stdout)
 	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
+		writer.WriteHeader(http.StatusInternalServerError)
 		_, _ = fmt.Fprintf(os.Stderr, "Error connecting to Kafka broker %q: %v\n", broker, err)
-		_, _ = fmt.Fprintf(w, "Error connecting to Kafka broker %q: %v\n", broker, err)
+		_, _ = fmt.Fprintf(writer, "Error connecting to Kafka broker %q: %v\n", broker, err)
 		return
-	} else {
-		defer func() {
-			if err := admin.Close() ; err != nil {
-				_, _ = fmt.Fprintf(os.Stderr, "Error disconnecting from Kafka broker %q: %v\n", broker, err)
-			}
-		}()
 	}
-
-	// NOTE: choice of underscore as separator is important as it is not allowed in k8s names
-	topicName := fmt.Sprintf("%s_%s", parts[0], parts[1])
-	topicDetail := sarama.TopicDetail{NumPartitions: 1, ReplicationFactor: 1}
-	if metadata, err := admin.DescribeTopics([]string{topicName}); err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		_, _ = fmt.Fprintf(os.Stderr, "Error trying to list topics to see if %q exists: %v\n", topicName, err)
-		_, _ = fmt.Fprintf(w, "Error trying to list topics to see if %q exists: %v\n", topicName, err)
-		return
-	} else if metadata[0].Err == sarama.ErrUnknownTopicOrPartition {
-		if err := admin.CreateTopic(topicName, &topicDetail, false); err != nil {
-			w.WriteHeader(http.StatusInternalServerError)
-			_, _ = fmt.Fprintf(os.Stderr, "Error creating topic %q: %v\n", topicName, err)
-			_, _ = fmt.Fprintf(w, "Error creating topic %q: %v\n", topicName, err)
-			return
+	defer func() {
+		if err := kafkaClient.Close(); err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "Error disconnecting from Kafka broker %q: %v\n", broker, err)
 		}
-		w.WriteHeader(http.StatusCreated)
-	} else if metadata[0].Err == sarama.ErrNoError {
-		w.WriteHeader(http.StatusOK)
-	} else {
-		w.WriteHeader(http.StatusInternalServerError)
-		_, _ = fmt.Fprintf(os.Stderr, "Error creating topic %q: %v\n", topicName, err)
-		_, _ = fmt.Fprintf(w, "Error creating topic %q: %v\n", topicName, err)
-		return
-	}
-
-	// Either created or already existed
-	w.Header().Set("Content-Type", "application/json")
-	res := result{Gateway: gateway, Topic: topicName}
-	if err := json.NewEncoder(w).Encode(res); err != nil {
-		_, _ = fmt.Fprintf(os.Stderr, "Failed to write json response: %v", err)
-		return
-	} else {
-		_, _ = fmt.Fprintf(os.Stderr, "Reported successful topic %q\n", topicName)
-	}
-}
-
-type result struct {
-	Gateway string `json:"gateway"`
-	Topic   string `json:"topic"`
+	}()
+	requestHandler := &handler.TopicCreationRequestHandler{KafkaClient: kafkaClient, Gateway: gateway, Writer: os.Stderr}
+	requestHandler.GetHandlerFunc()(writer, request)
```

Comment:
Why introduce a function that returns a function here?

Reply:
That way, it adheres to `HandlerFunc` and can be directly tested with `myHandler.ServeHTTP`

Reply:
I'll word it the other way then: why create a struct? It does not have any other method and just serves as a trampoline for creating the handler

Why not 
```
   NewHandler(KafkaClient, Gateway,Writer) HandlerFunc
```

I'll merge as-is for now, but I think we should revisit.